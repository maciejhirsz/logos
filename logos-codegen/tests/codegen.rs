use insta::assert_snapshot;
use std::{error::Error, path::PathBuf};
use syn::parse_file;

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
#[case("skip_callback")]
#[case("skip_callback_failure")]
#[case("error_callback0")]
#[case("error_callback1")]
#[case("error_callback_failure")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let rust_ver = if cfg!(rust_1_82) { "1_82" } else { "pre_1_82" };

    let codegen_alg = if cfg!(feature = "state_machine_codegen") {
        "state_machine"
    } else {
        "tailcall"
    };

    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push("codegen");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    let input = std::fs::read_to_string(input)?;

    let generated = logos_codegen::generate(input.parse()?).to_string();

    let formatted =
        prettyplease::unparse(&parse_file(&generated).expect("Logos output is unparseable"));

    assert_snapshot!(format!("{fixture}-{codegen_alg}-{rust_ver}"), formatted);

    Ok(())
}
