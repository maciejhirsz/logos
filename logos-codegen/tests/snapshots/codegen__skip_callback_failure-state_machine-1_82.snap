---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        fn _logos_derive_compile_errors() {
            {
                compile_error!(
                    "Expected #[logos(skip(\"regex literal\"[, [callback = ] callback, priority = priority]))]"
                )
            }
            {
                compile_error!(
                    "Expected a named argument at this position\n\nhint: If you are trying to define a callback here use: callback = ..."
                )
            }
            { compile_error!("Expected: priority = <integer>") }
            { compile_error!("Expected an unsigned integer") }
            { compile_error!("Resetting previously set priority") }
            {
                compile_error!(
                    "Inline callbacks must use closure syntax with exactly one parameter"
                )
            }
            { compile_error!("Callback has been already set") }
            { compile_error!("Previous callback set here") }
            { compile_error!("Expected: callback = ...") }
            {
                compile_error!(
                    "Unknown nested attribute: unknown\n\nExpected: callback or priority"
                )
            }
            {
                compile_error!(
                    "Expected: #[logos(skip \"regex literal\")] or #[logos(skip(...))]"
                )
            }
            { compile_error!("expected literal") }
            {
                compile_error!(
                    "Expected #[logos(skip(\"regex literal\"[, [callback = ] callback, priority = priority]))]"
                )
            }
            { compile_error!("Expected a &str or &[u8] slice") }
            { compile_error!("Invalid nested attribute") }
            {
                compile_error!(
                    "The following variants can all match simultaneously: `Leaf { pattern: Pattern { source: \"a\", hir: Literal(\"a\") }, span: Span, priority: 2, kind: Skip, callback: None }`, `Leaf { pattern: Pattern { source: \"a\", hir: Literal(\"a\") }, span: Span, priority: 2, kind: Skip, callback: Some(Inline(InlineCallback { arg: Ident(_), body: TokenStream [], span: Span })) }`, `Leaf { pattern: Pattern { source: \"a\", hir: Literal(\"a\") }, span: Span, priority: 2, kind: Unit(Ident(A)), callback: None }`"
                )
            }
        }
    }
}
